test/unit/cors_library.rb

```diff
@@ -0,0 +1,26 @@
+# frozen_string_literal: true
+
+# Copyright 2015-2017, the Linux Foundation, IDA, and the
+# CII Best Practices badge contributors
+# SPDX-License-Identifier: MIT
+
+require 'test_helper'
+
+class CorsLibraryTest < ActiveSupport::TestCase
+  test 'CORS library middleware stack location check in production' do
+    # Because of the way it works, Rack::Cors *must* be first in the Rack
+    # middleware stack, as documented here: https://github.com/cyu/rack-cors
+    # This test verifies this precondition, because it'd be easy to
+    # accidentally cause this assumption to fail as code is changed and
+    # gems are added or updated.
+    first = `RAILS_ENV=production rake middleware | grep '^use ' | head -1`
```

Comment:
This works because you're reassigning `first` rather than doing `first.chomp!`, but I think it's best practice to actually treat strings as immutable. So, I would consider:
```ruby
first = `RAILS_ENV=production rake middleware`.lines.grep(/^use /).first
```

Reply:
Actually, for clarity, I would refactor to a single test:
```ruby
  test 'CORS library middleware stack location check in environments' do
    %w(production test).each do |environment|
      command = "RAILS_ENV=${environment} rake middleware"
      result = %x(command).lines.grep(/^use /).first
      assert_equal 'use Rack::Cors', result
  end
```

Reply:
Thanks!! A few replies...

I *am* treating strings as immutable.  An expression like x = y.first doesn't mutate string y; instead it creates a new string and then makes x refer to that new string.  I believe you mean "treat variables as constants once assigned", right?  Which is fine, BTW, I've done lots of functional programming :-).

I originally didn't refactor this because using a loop means that you can't just use the line number to see which case failed.  And to be fair, I wrote one first, and did the second as an afterthought.  But failure shouldn't happen, and fixing it will require some diving anyway, so I think you're right about the refactor.

I think your proposed code has a bug; the command line uses "$" (the bash inliner) but that's a Ruby string so I think "#" should be used instead, resulting in:

> command = "RAILS_ENV=#{environment} rake middleware"

The "each" also needs a matching "end", but I can do that :-).

I wonder if we should test "development" as well as production and test - that would slow things even further, but it'd make sure that environment worked too. As always, you can always add more tests...!

Reply:
It turns out that %x (and backquotes) won't work in a loop, both expect a literal.